src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1243,10 +1243,116 @@ public void OnHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
             }
         }
 
+        public void OnStaticIndexedHeader(int index)
+        {
+            Debug.Assert(index <= H2StaticTable.Count);
+
+            ref readonly var entry = ref H2StaticTable.Get(index - 1);
+            OnStaticIndexedHeaderCore(index, entry.Name, entry.Value);
+        }
+
+        public void OnStaticIndexedHeader(int index, ReadOnlySpan<byte> value)
+        {
+            Debug.Assert(index <= H2StaticTable.Count);
+
+            OnStaticIndexedHeaderCore(index, H2StaticTable.Get(index - 1).Name, value);
+        }
+
+        // We can't throw a Http2StreamErrorException here, it interrupts the header decompression state and may corrupt subsequent header frames on other streams.
+        // For now these either need to be connection errors or BadRequests. If we want to downgrade any of them to stream errors later then we need to
+        // rework the flow so that the remaining headers are drained and the decompression state is maintained.
+        public void OnStaticIndexedHeaderCore(int index, ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
+        {
+            // https://tools.ietf.org/html/rfc7540#section-6.5.2
+            // "The value is based on the uncompressed size of header fields, including the length of the name and value in octets plus an overhead of 32 octets for each header field.";
+            _totalParsedHeaderSize += HeaderField.RfcOverhead + name.Length + value.Length;
+            if (_totalParsedHeaderSize > _context.ServiceContext.ServerOptions.Limits.MaxRequestHeadersTotalSize)
+            {
+                throw new Http2ConnectionErrorException(CoreStrings.BadRequest_HeadersExceedMaxTotalSize, Http2ErrorCode.PROTOCOL_ERROR);
```

Comment:
It does feel bad to kill the entire connection and with it any concurrent streams because the client exceeded Kestrel's MaxRequestHeadersTotalSize which isn't communicated via the protocol. Can we set some state (maybe the  Http2StreamErrorException) that tells us to continue decoding the headers, but not store them, and then rethrow the Http2StreamErrorException after we're done decoding?

Reply:
It is existing behavior from the regular header path. Create an issue for it.

Reply:
https://github.com/dotnet/aspnetcore/issues/17861